drivers/adc/adc_nrfx_saadc.c

```diff
@@ -432,11 +445,42 @@ static int check_buffer_size(const struct adc_sequence *sequence,
 	return 0;
 }
 
+static bool has_single_ended(const struct adc_sequence *sequence)
+{
+	return sequence->channels & m_data.single_ended_channels;
+}
+
+static void correct_single_ended(const struct adc_sequence *sequence,
+				 uint8_t active_channels)
+{
+	int8_t channel_id = -1;
+	uint32_t selected_channels = sequence->channels;
+	int16_t *sample_it = nrf_saadc_buffer_pointer_get(NRF_SAADC);
+	int16_t *sample_it_end = sample_it + active_channels;
+
+	do {
+		do {
+			channel_id = channel_id < (SAADC_CH_NUM - 1) ? channel_id + 1 : 0;
+		} while (!(selected_channels & BIT(channel_id)));
+
+		if (!(m_data.single_ended_channels & BIT(channel_id))) {
+			continue;
+		}
+
+		if (*sample_it >= 0) {
+			continue;
+		}
+
+		*sample_it = 0;
+	} while (++sample_it != sample_it_end);
```

Comment:
Couldn't this be simplified a bit? How about something like this:
```
	uint32_t channel_bit = BIT(0);
	int16_t *sample = nrf_saadc_buffer_pointer_get(NRF_SAADC);

	while (channel_bit < selected_channels) {
		if (selected_channels & channel_bit) {
			if ((single_ended_channels & channel_bit) && *sample < 0) {
				*sample = 0;
			}

			++sample;
		}

		channel_bit <<= 1;
	}
```
You wouldn't need then to pass `active_channels` to this function.

Reply:
Right, it was quite complex as I was iterating over every sampling in the sequence, I will clean it up :)